[Backend] Seed data

michellemtchai opened the pull request:
Please see issue #166 for more detail.

**Please merge this PR after PR #182**

To test this PR:
1) `docker-compose up --build`
2) `docker-compose exec api rake db:drop db:create db:migrate`
3) `docker-compose exec api rake db:seed`: There should be a log that tells you if the seeding succeeded or not
4) use browser to check db state via GET routes
6) change the content in `api/db/seed/mock_data.json`
5) `docker-compose exec api rake db:seed`: The db should update to match the changed `mock_data`
7) use browser to check db state via GET routes

**Note:** The last line in `seed.rb` generates fake data. You can comment it out if you only want to deal with the data from the `mock_data.json`.

gabriellesc commented:
Given that this is a hot fix, should there be a twin PR towards develop?

michellemtchai committed:
feat(db): setup request and readJson in seeds.rb
request: calls route in controller
readJson: read json file in db/seed

siefkenj commented:
Can you rebase this branch now that PR #152 has been merged?

michellemtchai commented:
Okay. I'll do that.

michellemtchai committed:
feat(api): setup seed data
create seeds_handler to handle api calls
create seed data for the following tables:
applicant
application
assignment
instructor
position
position_preference
position_template
reporting_tag
session
wage_chunk

michellemtchai committed:
fix(db): set create for multiple id to update automatically
make it so that when instructor is created, there is a trigger to update it so that the relation in position_ids is made

siefkenj commented on api/app/helpers/seeds_handler.rb:
```diff
@@ -0,0 +1,231 @@
+        end
+    end
+
+    def insert_data(seed)
```
I don't understand why you need all these functions...all your data should be seeded by making calls to the api with `request`. You should never need to peek inside any controllers...

michellemtchai replied:
So, we're chaining request outcome to make more requests?

siefkenj commented on api/db/seeds.rb:
```diff
@@ -3,65 +3,150 @@
+        update: '/reporting_tags/:id',
+        file: 'reporting_tag.json',
+        unique: [:wage_chunk_id, :position_id],
+        foreign_keys: [
```
Why do you need to know anything about foreign keys?

michellemtchai replied:
I actually used it to get the actually id value like in the discussion below.

siefkenj commented on api/db/seed/wage_chunk.json:
```diff
@@ -0,0 +1,65 @@
+        "rate": 42.25,
+        "start_date": "2018-9-10",
+        "end_date": "2018-12-10",
+        "assignment_id": 0
```
The ids shouldn't be hardcoded...they should be read from the return value of the api call and synthesized on the spot. Eventually the controller will generate randomized ids, and then this seeing will fail.

michellemtchai replied:
Actually, this only looks like `id`s it's actually the index of entry from `assignment.json`.

siefkenj replied:
Oh. That's pretty confusing then...If it's separate linking data, put it in a separate JSON object. That way the seed data looks like seed data.

michellemtchai replied:
Do you mean like in a separate file? I'm having trouble imagining how that would work.

siefkenj replied:
Yes. Maybe `wage_chunk_links.json`. You're essentially creating a DSL for batch-processing with the API right now...

michellemtchai replied:
Is the DSL comparison what I'm doing right now or what I should/should not be attempting? Also, the seeding should be done via chaining API in order(?)

siefkenj replied:
@michellemtchai Exactly! You're chaining the API calls. If you want to create a mini DSL to assist you with chaining a bunch of calls in a particular order, that's fine. If you want to hard code it, that's also fine.

Whatever sequence of calls you code should also be able to be ported more-or-less directly to Javascript for client-side unit tests.

michellemtchai replied:
Okay. I think I get it now. Thanks.

michellemtchai committed:
feat(db): setup fake data
use FakeData module to setup fake data for each module
update gem for Fake

michellemtchai committed:
feat(db): set up mock json for seed
create mock json data for seeding
update fake_data to match format

michellemtchai committed:
feat(api): link fake_data to seed_handler
setup structure to generate fake data using seed_handler
remove global variable from fake_data functions

michellemtchai committed:
feat(api): setup structure for chaining
setup data structure for chaining the routes together

michellemtchai committed:
feat(api): set seed handler to get current db state
get each table in db before starting seeding process

michellemtchai committed:
feat(api): setup structure for post request

michellemtchai committed:
feat(db): finish insert_data for seed_handler
finish seed_data insertion function

dingtaoliu commented on api/app/controllers/api/v1/applicants_controller.rb:
```diff
@@ -34,6 +39,17 @@ def update
+        def delete
+            params.require(:id)
+            applicant = Applicant.find(params[:id])
+            if session.destroy!
```
Should this be `applicant.destroy!`?

michellemtchai replied:
Yes. This is a mistake that carried over from PR #183. Once, I fix that PR, I'll rebase this branch off that PR, again.

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -16,7 +16,14 @@ def index
+
+            params.require(:applicant_id)
             if invalid_id(Session, :session_id) then return end
+            if invalid_id(Applicant, :applicant_id) then return end
```
you can write this as `return if invalid_id(Applicant, :applicant_id)` so its a little easier to read.

michellemtchai replied:
Okay. I'll do that. Though, do you mind commenting on PR #183 instead, if you're going to review the controller.

michellemtchai replied:
Done.

dingtaoliu commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -84,13 +107,21 @@ def matching_data_update_params
+        def all_applications
+            exclusion = [:id, :created_at, :updated_at, :application_id]
+            return Application.order(:id).map do |entry|
+                matching = ApplicantDataForMatching.find_by(application_id: entry[:id])
```
it might be better to use a join here so we don't need to make a db call for each Application, but we don't need to worry about it for now.

dingtaoliu commented on api/app/controllers/api/v1/assignments_controller.rb:
```diff
@@ -44,6 +49,17 @@ def update
+        def delete
+            params.require(:id)
+            assignment = Assignment.find(params[:id])
+            if session.destroy!
```
should this be `assignment.destroy!`?

michellemtchai replied:
I've fixed it.

dingtaoliu commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -14,30 +14,44 @@ def index
+        def delete
+            params.require(:id)
+            instructor = Instructor.find(params[:id])
+            if session.destroy!
```
this might be a type as well

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -38,6 +48,17 @@ def update
+        def delete
+            params.require(:id)
+            position_template = PostionTemplate.find(params[:id])
+            if session.destroy!
```
typo

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -129,11 +145,35 @@ def matching_update_params
+        end
+
+        def update_instructors_ids(position)
+            if params[:instructor_ids]
```
you should use a guard clause for these kinds of nested conditionals, `return if params[:instructor_ids].blank?`

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/reporting_tags_controller.rb:
```diff
@@ -29,6 +39,17 @@ def update
+        def delete
+            params.require(:id)
+            reporting_tag = ReportingTag.find(params[:id])
+            if session.destroy!
```
typo

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/wage_chunks_controller.rb:
```diff
@@ -36,6 +41,17 @@ def update
+        def delete
+            params.require(:id)
+            wage_chunk = WageChunk.find(params[:id])
+            if session.destroy!
```
typo

michellemtchai replied:
Fixed.

dingtaoliu commented on api/app/controllers/api/v1/sessions_controller.rb:
```diff
@@ -39,12 +39,13 @@ def update
-                end
+            params.require(:id)
+            session = Session.find(params[:id])
+            if session.destroy!
```
not typo!

dingtaoliu commented on api/app/helpers/application_helper.rb:
```diff
@@ -2,4 +2,18 @@
 # Base class from which most applications use
 module ApplicationHelper
+	def json(data, include: nil, except: nil)
+		data = JSON.parse(data.to_json.to_s, symbolize_names: true)
```
if you just want the keys in data to be symbolized, you can do `data = data.with_indifferent_access` instead of converting it into a json string and then parsing the json.

dingtaoliu commented on api/db/seeds.rb:
```diff
@@ -3,65 +3,73 @@
-  User.create!(row.with_indifferent_access)
-end
+insert_data(chaining, file: 'mock_data.json')
+insert_data(chaining, entries: entries)
```
let's comment out the auto generated data for now.

michellemtchai replied:
Actually, I plan to remove this so that it's a separate task.

dingtaoliu commented on api/app/helpers/seeds_handler.rb:
```diff
@@ -0,0 +1,184 @@
+        return route.scan(/(\/add_)/).length > 0
+    end
+
+    def get_route(route, data)
```
can we add the name of the route into chaining? This avoids potential parsing errors.

michellemtchai replied:
Actually, I can't because I don't know the value of `:session_id` or `:id`, so I can't include them.

michellemtchai committed:
feat(db): remove update portion of seed

dingtaoliu commented:
I'm good with these changes, I think it can be merged.  Is it okay that the unit tests failed?

siefkenj commented:
@dingtaoliu We don't merge PRs with broken tests. This PR also needs to wait for  #183 to be ready.

michellemtchai committed:
fix(db): change available_position_template result to match frontend
use [{position_template: filename}] as opposed to [filename]

siefkenj commented on api/db/seeds.rb:
```diff
@@ -3,65 +3,72 @@
+include SeedsHandler
 
-seed_data_dir = Rails.root.join('db', 'seed')
+chaining = [
```
variables like this should be named as nouns, not verbs. Something like `seed_data_sequence` would be better.

michellemtchai replied:
Fixed.

siefkenj commented on api/db/seeds.rb:
```diff
@@ -3,65 +3,72 @@
+    {
+        get: '/sessions',
+        create: '/sessions',
+        unique: [:name]
```
`unique` is a confusing name. What do you think of `index_on`?

michellemtchai replied:
Fixed.

siefkenj commented on api/app/helpers/seeds_handler.rb:
```diff
@@ -0,0 +1,259 @@
+                log.push('No JSON file was given.')
+            end
+        end
+        puts log
```
Isn't there a more offical way to log something?

michellemtchai replied:
There is. However, the problem with that it is the logging goes immediately into the log of the rest of the stuff Rack is logging at the same time, so any message that's relevant to the user kind of gets lost in between other really long messages. This is why I chose to accumulate the log messages because I print them.

siefkenj commented on api/app/helpers/seeds_handler.rb:
```diff
@@ -0,0 +1,259 @@
+        return true, log
+    end
+
+    def post_data(chaining, json, log)
```
You're passing around `log` between several functions. This is a pattern `class`es were invented to solve. Would it make sense to refactor some of this into a class?

michellemtchai replied:
I'm not sure which pattern class you're referring to. For now, I made use of `@log` to make it so that there's no passing around `log` through several functions.

michellemtchai committed:
fix(api): change variable names to more descriptive ones

michellemtchai committed:
fix(api): remove log as a function parameter in seed_handler
change to using @log
